electrolytes/solvation_shell_utils.py

```diff
@@ -0,0 +1,152 @@
+    """
+    actual_shells = [st.extract(shell, copy_props=True) for shell in shells]
+    filtered_shells = [
+        shell
+        for shell in actual_shells
+        if not any(
+            at.getResidue().pdbres.strip() in solute_res_names for at in shell.atom
+        )
+    ]
+    return filtered_shells
+
+
+def expand_shell(
+    st: Structure,
+    shell_ats: Set[int],
+    central_solute: int,
+    radius: float,
+    solute_res_names: List[str],
+    max_shell_size: int = 200,
+) -> Structure:
+    """
+    Expands a solvation shell. If there are any (non-central) solutes present in the shell,
+    recursively include shells around those solutes.
+    First, gets the molecule numbers of solute molecules that are within the radius
+    and not already expanded around. Then, continuously expand around them as long as we don't hit an atom limit.
+    Args:
+        st: Entire structure from the PDB file
+        shell_ats: Set of atoms in a shell (1-indexed)
+        central_solute: Index of the central solute atom in the shell
+        radius: Solvation radius (Angstroms) to consider
+        solute_res_names: List of residue names that correspond to solute atoms in the simulation
+        max_shell_size: Maximum size (in atoms) of the expanded shell
+    Returns:
+        Structure object containing the expanded solvation shell
+    """
+    # TODO: I'm not sure why we have to do this, but sometimes the central solute is not in the shell
+    if central_solute not in shell_ats:
+        shell_ats.add(central_solute)
+
+    solutes_included = set([central_solute])
+    while len(shell_ats) < max_shell_size:
+        new_solutes = set()
+        for at in shell_ats:
+            # atom is part of a non-central solute molecule - should expand the shell
+            if (
+                st.atom[at].molecule_number not in solutes_included
+                and st.atom[at].getResidue().pdbres.strip() in solute_res_names
+            ):
+                new_solutes.add(st.atom[at].molecule_number)
+
+        if new_solutes:
+            # add entire residues within solvation shell radius of any extra solute atoms
+            shell_ats.update(
+                analyze.evaluate_asl(
+                    st,
+                    f'fillres within {radius} mol {",".join([str(i) for i in new_solutes])}',
+                )
+            )
+            solutes_included.update(new_solutes)
+        else:
+            break
+    shell_ats = sorted(shell_ats)
+    final_structure = st.extract(shell_ats, copy_props=True)
+
+    # find index of central solute in the sorted shell_ats (adjust for 1-indexing)
+    central_solute_idx = shell_ats.index(central_solute) + 1
+
+    # contract everthing to be centered on our molecule of interest
+    # (this will also handle if a molecule is split across a PBC)
+
+    clusterstruct.contract_structure2(
+        final_structure, contract_on_atoms=[central_solute_idx]
```

Comment:
```suggestion
        final_structure, contract_on_atoms=st.molecule[central_solute].getAtomList()[0:1]
```

Ah, you just want one atom of the central solute. This will work.

Reply:
Ah thanks, I figured there was a better way to do this!

Reply:
Your suggestion doesn't quite work because the numbering has to be in accordance with `final_structure` (which was constructed with `shell_ats` ), not the original `st` which has more atoms. See my fix and let me know if makes sense.